instawow/resolvers.py

```diff
@@ -125,6 +129,33 @@ def decompose_url(cls, uri: str) -> Optional[Tuple[str, str]]:
                     version=file['displayName'],
                     options=PkgOptions(strategy=strategy.name))
 
+    async def search(self, search_pattern: str):
+        from lxml.html import document_fromstring
+
+        search_url = self.curse_url / 'search'
+        params = {'search': search_pattern}
+        async with self.web_client.get(search_url, params=params) as response:
+            page_text = await response.text()
+
+            if "No results for" in page_text:
+                return []
+
+            html = document_fromstring(page_text)
+            rows = html.xpath(
+                '//div[contains(@class, "project-listing-row")]')
+            addons = []
+            for row in rows:
+                name = row.xpath('.//a/h3')[0].text
+                path = (row.xpath('.//a[contains(@href, "wow/addons")]')[0]
+                        .attrib.get('href').lstrip('/'))
+                url = URL('https://www.curseforge.com') / path
+                author = (row.xpath('.//a[contains(@href, "members/")]')[0]
+                          .text)
+                desc = (row.xpath('.//p')[0].text).strip()
+                addons.append({'name': name, 'url': str(url), 'author': author,
```

Comment:
You could create a `Pkg_` instead of makeshift `dict`s here (all the necessary data is available through the search API) or delegate looking up the package to `CurseResolver.resolve` to keep things DRY.

Reply:
Right, but this would require `O(n)` queries I think, so I'd like to avoid that. I'll look at whether it's possible to use Pkg_ directly, that was actually my initial intent but it was time to sleep.